libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/MapboxRouteOptionsProvider.kt

```diff
@@ -1,21 +1,47 @@
-package com.mapbox.navigation.core.directions.session
+package com.mapbox.navigation.core.routeoptions
 
 import android.location.Location
 import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.base.common.logger.Logger
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
 import com.mapbox.geojson.Point
 import com.mapbox.navigation.base.trip.model.RouteProgress
-import com.mapbox.navigation.core.trip.session.TripSession
 
-internal object AdjustedRouteOptionsProvider {
+/**
+ * Default implementation of [RouteOptionsProvider]
+ */
+internal class MapboxRouteOptionsProvider(
+    private val logger: Logger
+) : RouteOptionsProvider {
 
-    private const val DEFAULT_REROUTE_BEARING_TOLERANCE = 90.0
+    private companion object {
+        const val DEFAULT_REROUTE_BEARING_TOLERANCE = 90.0
+    }
 
-    fun getRouteOptions(directionsSession: DirectionsSession, tripSession: TripSession, location: Location): RouteOptions? {
-        val routeOptions: RouteOptions = directionsSession.getRouteOptions() ?: return null
-        val routeProgress: RouteProgress = tripSession.getRouteProgress() ?: return null
+    /**
+     * Provides a new instance of *RouteOptions* based on initial *RouteOptions*, *RouteProgress* and
+     * current *Location*
+     *
+     * Returns *null* if a new [RouteOptions] instance cannot be combined based on the input given.
+     */
+    override fun update(
+        routeOptions: RouteOptions?,
+        routeProgress: RouteProgress?,
+        location: Location?
+    ): RouteOptions? {
+        if (routeOptions == null || routeProgress == null || location == null) {
+            logger.w(
+                Tag("MapboxRouteOptionsProvider"),
+                Message("Cannot combine RouteOptions, invalid inputs. *routeOptions*, " +
+                    "*routeProgress*, and *location* mustn't be null")
+            )
+            return null
```

Comment:
Can we make the `update` method return not-null and throw a custom exception if the combination fails? We could catch that exception when invoking an properly report using `RerouteState#Failed(msg, exception)` instead of logging.

Reply:
I see how it helps, it's reasonable. But I'm forcing to avoid exception driven architecture. What do you think we we return `sealed class`?

Reply:
We can return an expected construct instead. Something along the lines of:
```
class RouteOptionsResult(val expected: SomeValue?, val error: SomeError?) {
  fun hasValue() = ...
  fun hasError() = ...
}
```

Reply:
it's an option, but not look as nice as sealed: 

Regular class, there're both values are nullable
```
class RouteOptionsResult1 {
        val routeOptions: RouteOptions?
        val error: Throwable?

        val hasError: Boolean
            get() = error != null

        val hasRouteOptions: Boolean
            get() = routeOptions != null

        constructor(routeOptions: RouteOptions) {
            this.routeOptions = routeOptions
            this.error = null
        }

        constructor(error: Throwable) {
            this.routeOptions = null
            this.error = error
        }
    }
```

Sealed class:
```
sealed class RouteOptionsResult2 {
        data class Success(val routeOptions: RouteOptions) : RouteOptionsResult2()
        data class Error(val error: Throwable) : RouteOptionsResult2()
    }
```

from java developers perspective a `sealed class` is just an abstract class that has `N`-number inheritors

Reply:
That works as well @RingerJK :+1: